commit ed31fc445fc6d4769b7bd119770ae1da9b15d2bf
Author: Endi Sukma Dewata <edewata@redhat.com>
Date:   Tue Feb 21 22:22:16 2012 -0600

    Replaced BtoA/AtoB with Apache codec.
    
    The OSUtil's BtoA() and AtoB() have been replaced by Base64
    codec from Apache Commons library. The codec is configured to
    use 64-byte line width as defined in RFC 1421.
    
    Ticket #90

--- a/pki/.classpath
+++ b/pki/.classpath
@@ -30,5 +30,6 @@
 	<classpathentry kind="lib" path="/usr/share/java/servlet.jar"/>
 	<classpathentry kind="lib" path="/usr/share/java/xerces-j2.jar"/>
 	<classpathentry kind="lib" path="/usr/share/java/junit4.jar"/>
+	<classpathentry kind="lib" path="/usr/share/java/commons-codec.jar"/>
 	<classpathentry kind="output" path="build/classes"/>
 </classpath>
--- a/pki/base/ca/src/CMakeLists.txt
+++ b/pki/base/ca/src/CMakeLists.txt
@@ -24,6 +24,13 @@
         ${JAVA_LIB_INSTALL_DIR}
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 
 # identify java sources
 set(pki-ca_java_SRCS
@@ -42,7 +49,7 @@
     ${PKI_CERTSRV_JAR} ${PKI_CMS_JAR} ${PKI_CMSCORE_JAR}
     ${PKI_CMSUTIL_JAR} ${PKI_NSUTIL_JAR}
     ${LDAPJDK_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR} ${SYMKEY_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR} ${SYMKEY_JAR})
 
 
 # set version
--- a/pki/base/common/src/CMakeLists.txt
+++ b/pki/base/common/src/CMakeLists.txt
@@ -24,6 +24,13 @@
         /usr/share/java
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 find_file(SERVLET_JAR
     NAMES
         servlet.jar
@@ -996,7 +1003,7 @@
 set(CMAKE_JAVA_INCLUDE_PATH
     ${PKI_NSUTIL_JAR} ${PKI_CMSUTIL_JAR}
     ${LDAPJDK_JAR} ${SERVLET_JAR} ${VELOCITY_JAR} ${XALAN_JAR} ${XERCES_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR} ${SYMKEY_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR} ${SYMKEY_JAR})
 
 set(CMAKE_JAVA_TARGET_VERSION ${APPLICATION_VERSION})
 
--- a/pki/base/console/src/CMakeLists.txt
+++ b/pki/base/console/src/CMakeLists.txt
@@ -77,6 +77,13 @@
         /usr/share/java
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 
 # identify java sources
 set(pki-console_java_SRCS
@@ -648,7 +655,8 @@
     ${MMC_EN_JAR} ${NMCLF_JAR} ${NMCLF_EN_JAR}
     ${PKI_NSUTIL_JAR}
     ${JSS_JAR}
-    ${OSUTIL_JAR})
+    ${OSUTIL_JAR}
+    ${COMMONS_CODEC_JAR})
 
 
 # set version
--- a/pki/base/java-tools/src/CMakeLists.txt
+++ b/pki/base/java-tools/src/CMakeLists.txt
@@ -16,6 +16,13 @@
         /usr/share/java
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 find_file(XALAN_JAR
     NAMES
         xalan-j2.jar
@@ -60,7 +67,7 @@
 set(CMAKE_JAVA_INCLUDE_PATH
     ${PKI_NSUTIL_JAR} ${PKI_CMSUTIL_JAR}
     ${XALAN_JAR} ${XERCES_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR})
 
 set(CMAKE_JAVA_TARGET_VERSION ${APPLICATION_VERSION})
 
--- a/pki/base/java-tools/templates/pki_java_command_wrapper.in
+++ b/pki/base/java-tools/templates/pki_java_command_wrapper.in
@@ -142,6 +142,7 @@
     # Fedora 16+
     CP=/usr/lib64/java/osutil.jar:${CP}
 fi
+CP=/usr/share/java/commons-codec.jar:${CP}
 CP=/usr/share/java/ldapjdk.jar:${CP}
 CP=/usr/share/java/${PRODUCT}/pki-nsutil.jar:${CP}
 CP=/usr/share/java/${PRODUCT}/pki-cmsutil.jar:${CP}
--- a/pki/base/java-tools/templates/pretty_print_cert_command_wrapper.in
+++ b/pki/base/java-tools/templates/pretty_print_cert_command_wrapper.in
@@ -142,6 +142,7 @@
     # Fedora 16+
     CP=/usr/lib64/java/osutil.jar:${CP}
 fi
+CP=/usr/share/java/commons-codec.jar:${CP}
 CP=/usr/share/java/ldapjdk.jar:${CP}
 CP=/usr/share/java/${PRODUCT}/pki-nsutil.jar:${CP}
 CP=/usr/share/java/${PRODUCT}/pki-cmsutil.jar:${CP}
--- a/pki/base/java-tools/templates/pretty_print_crl_command_wrapper.in
+++ b/pki/base/java-tools/templates/pretty_print_crl_command_wrapper.in
@@ -142,6 +142,7 @@
     # Fedora 16+
     CP=/usr/lib64/java/osutil.jar:${CP}
 fi
+CP=/usr/share/java/commons-codec.jar:${CP}
 CP=/usr/share/java/ldapjdk.jar:${CP}
 CP=/usr/share/java/${PRODUCT}/pki-nsutil.jar:${CP}
 CP=/usr/share/java/${PRODUCT}/pki-cmsutil.jar:${CP}
--- a/pki/base/kra/src/CMakeLists.txt
+++ b/pki/base/kra/src/CMakeLists.txt
@@ -62,6 +62,13 @@
         ${JAVA_LIB_INSTALL_DIR}
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 find_file(SYMKEY_JAR
     NAMES
         symkey.jar
@@ -91,7 +98,7 @@
     ${PKI_CERTSRV_JAR} ${PKI_CMS_JAR} ${PKI_CMSCORE_JAR}
     ${PKI_CMSUTIL_JAR} ${PKI_NSUTIL_JAR}
     ${LDAPJDK_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR} ${SYMKEY_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR} ${SYMKEY_JAR})
 
 
 # set version
--- a/pki/base/ocsp/src/CMakeLists.txt
+++ b/pki/base/ocsp/src/CMakeLists.txt
@@ -62,6 +62,13 @@
         ${JAVA_LIB_INSTALL_DIR}
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 find_file(SYMKEY_JAR
     NAMES
         symkey.jar
@@ -84,7 +91,7 @@
     ${PKI_CERTSRV_JAR} ${PKI_CMS_JAR} ${PKI_CMSCORE_JAR}
     ${PKI_CMSUTIL_JAR} ${PKI_NSUTIL_JAR}
     ${LDAPJDK_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR} ${SYMKEY_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR} ${SYMKEY_JAR})
 
 
 # set version
--- a/pki/base/setup/pkicreate
+++ b/pki/base/setup/pkicreate
@@ -181,6 +181,7 @@
 my $ldapjdk_jar_base_name         = "ldapjdk.jar";         # CA, KRA, OCSP, TKS
 my $pki_nsutil_jar_base_name      = "pki-nsutil.jar";      # CA, KRA, OCSP, TKS
 my $osutil_jar_base_name          = "osutil.jar";          # CA, KRA, OCSP, TKS
+my $commons_codec_jar_base_name   = "commons-codec.jar";   # CA, KRA, OCSP, TKS
 my $symkey_jar_base_name          = "symkey.jar";          # CA, KRA, OCSP, TKS
 my $tomcatjss_jar_base_name       = "tomcatjss.jar";       # CA, KRA, OCSP, TKS
 my $velocity_jar_base_name        = "velocity.jar";        # CA, KRA, OCSP, TKS
@@ -479,6 +480,8 @@
 my $pki_nsutil_jar_symlink_path                  = undef;  # CA, KRA, OCSP, TKS
 my $osutil_jar_file_path                         = undef;  # CA, KRA, OCSP, TKS
 my $osutil_jar_symlink_path                      = undef;  # CA, KRA, OCSP, TKS
+my $commons_codec_jar_file_path                  = undef;  # CA, KRA, OCSP, TKS
+my $commons_codec_jar_symlink_path               = undef;  # CA, KRA, OCSP, TKS
 my $symkey_jar_file_path                         = undef;  # CA, KRA, OCSP, TKS
 my $symkey_jar_symlink_path                      = undef;  # CA, KRA, OCSP, TKS
 my $tomcatjss_jar_file_path                      = undef;  # CA, KRA, OCSP, TKS
@@ -1802,6 +1805,11 @@
             return 0;
         }
 
+        if (!defined($commons_codec_jar_file_path = find_jar($commons_codec_jar_base_name))) {
+            emit("could not find jar: $commons_codec_jar_base_name", "error");
+            return 0;
+        }
+
         if (!defined($pki_subsystem_jar_file_path = find_jar($pki_subsystem_jar_base_name))) {
             emit("could not find jar: $pki_subsystem_jar_base_name", "error");
             return 0;
@@ -1875,6 +1883,8 @@
                                            . "/" . $pki_nsutil_jar_base_name;
         $osutil_jar_symlink_path           = $webinf_lib_instance_path
                                            . "/" . $osutil_jar_base_name;
+        $commons_codec_jar_symlink_path    = $webinf_lib_instance_path
+                                           . "/" . $commons_codec_jar_base_name;
         $symkey_jar_symlink_path           = $webinf_lib_instance_path
                                            . "/" . $symkey_jar_base_name;
         $pki_subsystem_jar_symlink_path    = $webinf_lib_instance_path
@@ -2869,6 +2879,10 @@
         return 0 if !create_symlink($osutil_jar_symlink_path, $osutil_jar_file_path,
                                     $pki_user, $pki_group);
 
+        # create instance symlink to "commons_codec.jar"
+        return 0 if !create_symlink($commons_codec_jar_symlink_path, $commons_codec_jar_file_path,
+                                    $pki_user, $pki_group);
+
         # create instance symlink to "${subsystem_type}.jar"
         return 0 if !create_symlink($pki_subsystem_jar_symlink_path, $pki_subsystem_jar_file_path,
                                     $pki_user, $pki_group);
--- a/pki/base/silent/scripts/pkisilent
+++ b/pki/base/silent/scripts/pkisilent
@@ -83,6 +83,7 @@
                 . "/usr/share/java/${PRODUCT}/pki-cmsutil.jar:"
                 . "/usr/share/java/${PRODUCT}/pki-tools.jar:"
                 . "/usr/share/java/ldapjdk.jar:"
+                . "/usr/share/java/commons-codec.jar:"
                 . "/usr/share/java/xerces-j2.jar:"
                 . "/usr/share/java/xml-commons-apis.jar:"
                 . "/usr/share/java/xml-commons-resolver.jar:";
--- a/pki/base/silent/src/CMakeLists.txt
+++ b/pki/base/silent/src/CMakeLists.txt
@@ -76,7 +76,7 @@
     ${PKI_CERTSRV_JAR} ${PKI_CMS_JAR}
     ${PKI_CMSUTIL_JAR} ${PKI_NSUTIL_JAR}
     ${LDAPJDK_JAR} ${XERCES_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR} ${SYMKEY_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR} ${SYMKEY_JAR})
 
 set(CMAKE_JAVA_TARGET_VERSION ${APPLICATION_VERSION})
 
--- a/pki/base/tks/src/CMakeLists.txt
+++ b/pki/base/tks/src/CMakeLists.txt
@@ -62,6 +62,13 @@
         ${JAVA_LIB_INSTALL_DIR}
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 find_file(SYMKEY_JAR
     NAMES
         symkey.jar
@@ -81,7 +88,7 @@
     ${PKI_CERTSRV_JAR} ${PKI_CMS_JAR} ${PKI_CMSCORE_JAR}
     ${PKI_CMSUTIL_JAR} ${PKI_NSUTIL_JAR}
     ${LDAPJDK_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR} ${SYMKEY_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR} ${SYMKEY_JAR})
 
 
 # set version
--- a/pki/base/util/src/CMakeLists.txt
+++ b/pki/base/util/src/CMakeLists.txt
@@ -24,6 +24,13 @@
         /usr/share/java
 )
 
+find_file(COMMONS_CODEC_JAR
+    NAMES
+        commons-codec.jar
+    PATHS
+        /usr/share/java
+)
+
 find_file(XALAN_JAR
     NAMES
         xalan-j2.jar
@@ -331,7 +338,7 @@
 
 set(CMAKE_JAVA_INCLUDE_PATH
     ${LDAPJDK_JAR} ${XALAN_JAR} ${XERCES_JAR}
-    ${JSS_JAR} ${OSUTIL_JAR})
+    ${JSS_JAR} ${OSUTIL_JAR} ${COMMONS_CODEC_JAR})
 
 set(CMAKE_JAVA_TARGET_VERSION ${APPLICATION_VERSION})
 
--- a/pki/base/util/src/com/netscape/cmsutil/util/Utils.java
+++ b/pki/base/util/src/com/netscape/cmsutil/util/Utils.java
@@ -17,6 +17,7 @@
 // --- END COPYRIGHT BLOCK ---
 package com.netscape.cmsutil.util;
 
+import org.apache.commons.codec.binary.Base64;
 
 import java.net.*;
 import java.io.*;
@@ -253,12 +254,12 @@
     }
 
     public static String base64encode(byte[] bytes) {
-        String string = com.netscape.osutil.OSUtil.BtoA(bytes);
+        String string = new Base64(64).encodeToString(bytes);
         return string;
     }
 
     public static byte[] base64decode(String string) {
-        byte[] bytes = com.netscape.osutil.OSUtil.AtoB(string);
+        byte[] bytes = Base64.decodeBase64(string);
         return bytes;
     }
 }
--- a/pki/scripts/prepare_dogtag_pki
+++ b/pki/scripts/prepare_dogtag_pki
@@ -211,6 +211,7 @@
     APACHE_COMMONS_LANG="jakarta-commons-lang"
     APACHE_COMMONS_LOGGING="jakarta-commons-logging"
 fi
+APACHE_COMMONS_CODEC="apache-commons-codec"
 JAKARTA_COMMONS_COLLECTIONS="jakarta-commons-collections"
 JAKARTA_COMMONS_DBCP="jakarta-commons-dbcp"
 JAKARTA_COMMONS_POOL="jakarta-commons-pool"
@@ -293,7 +294,7 @@
 XML_COMMONS_RESOLVER="xml-commons-resolver"
 
 # Create a catch-all variable for PKI Support Packages
-PKI_SUPPORT_PACKAGES="${APR} ${APR_DEVEL} ${APR_UTIL} ${APR_UTIL_DEVEL} ${EXPAT} ${EXPAT_DEVEL} ${HTTPD} ${HTTPD_DEVEL} ${HTTPD_TOOLS} ${PCRE} ${PCRE_DEVEL} ${TOMCAT6} ${TOMCAT6_LIB} ${APACHE_COMMONS_LANG} ${APACHE_COMMONS_LANG} ${APACHE_COMMONS_LOGGING} ${JAKARTA_COMMONS_COLLECTIONS} ${JAKARTA_COMMONS_DBCP} ${JAKARTA_COMMONS_POOL} ${NSPR} ${NSPR_DEVEL} ${NSS} ${NSS_DEVEL} ${NSS_TOOLS} ${JSS} ${JSS_JAVADOC} ${TOMCATJSS} ${MOD_NSS} ${MOD_PERL} ${MOD_REVOCATOR} ${IDM_CONSOLE_FRAMEWORK} ${CYRUS_SASL} ${CYRUS_SASL_DEVEL} ${LDAPJDK} ${OPENLDAP} ${OPENLDAP_CLIENTS} ${OPENLDAP_DEVEL} ${OSUTIL} ${PERL_CRYPT_SSLEAY} ${PERL_DBD_SQLITE} ${PERL_DBI} ${PERL_HTML_PARSER} ${PERL_HTML_TAGSET} ${PERL_LIBWWW_PERL} ${PERL_MOZILLA_LDAP} ${PERL_PARSE_RECDESCENT} ${PERL_URI} ${PERL_XML_NAMESPACESUPPORT} ${PERL_XML_PARSER} ${PERL_XML_SAX} ${PERL_XML_SIMPLE} ${ESC} ${SVRCORE} ${SVRCORE_DEVEL} ${POLICYCOREUTILS} ${SELINUX_POLICY_DEVEL} ${SELINUX_POLICY_TARGETED} ${SQLITE} ${SQLITE_DEVEL} ${VELOCITY} ${BCEL} ${JAKARTA_ORO} ${JDOM} ${LOG4J} ${REGEXP} ${WERKEN_XPATH} ${XALAN_J2} ${XERCES_J2} ${XML_COMMONS_APIS} ${XML_COMMONS_RESOLVER}"
+PKI_SUPPORT_PACKAGES="${APR} ${APR_DEVEL} ${APR_UTIL} ${APR_UTIL_DEVEL} ${EXPAT} ${EXPAT_DEVEL} ${HTTPD} ${HTTPD_DEVEL} ${HTTPD_TOOLS} ${PCRE} ${PCRE_DEVEL} ${TOMCAT6} ${TOMCAT6_LIB} ${APACHE_COMMONS_LANG} ${APACHE_COMMONS_LANG} ${APACHE_COMMONS_LOGGING} ${APACHE_COMMONS_CODEC} ${JAKARTA_COMMONS_COLLECTIONS} ${JAKARTA_COMMONS_DBCP} ${JAKARTA_COMMONS_POOL} ${NSPR} ${NSPR_DEVEL} ${NSS} ${NSS_DEVEL} ${NSS_TOOLS} ${JSS} ${JSS_JAVADOC} ${TOMCATJSS} ${MOD_NSS} ${MOD_PERL} ${MOD_REVOCATOR} ${IDM_CONSOLE_FRAMEWORK} ${CYRUS_SASL} ${CYRUS_SASL_DEVEL} ${LDAPJDK} ${OPENLDAP} ${OPENLDAP_CLIENTS} ${OPENLDAP_DEVEL} ${OSUTIL} ${PERL_CRYPT_SSLEAY} ${PERL_DBD_SQLITE} ${PERL_DBI} ${PERL_HTML_PARSER} ${PERL_HTML_TAGSET} ${PERL_LIBWWW_PERL} ${PERL_MOZILLA_LDAP} ${PERL_PARSE_RECDESCENT} ${PERL_URI} ${PERL_XML_NAMESPACESUPPORT} ${PERL_XML_PARSER} ${PERL_XML_SAX} ${PERL_XML_SIMPLE} ${ESC} ${SVRCORE} ${SVRCORE_DEVEL} ${POLICYCOREUTILS} ${SELINUX_POLICY_DEVEL} ${SELINUX_POLICY_TARGETED} ${SQLITE} ${SQLITE_DEVEL} ${VELOCITY} ${BCEL} ${JAKARTA_ORO} ${JDOM} ${LOG4J} ${REGEXP} ${WERKEN_XPATH} ${XALAN_J2} ${XERCES_J2} ${XML_COMMONS_APIS} ${XML_COMMONS_RESOLVER}"
 
 ###########################################
 # Establish PKI Installation Dependencies #
--- a/pki/specs/dogtag-pki.spec
+++ b/pki/specs/dogtag-pki.spec
@@ -68,6 +68,8 @@
 %endif
 %endif
 
+Requires:         apache-commons-codec
+
 # Make certain that this 'meta' package requires the latest version(s)
 # of ALL top-level Dogtag PKI support packages
 Requires:         jss >= %{jss_version}
--- a/pki/specs/pki-core.spec
+++ b/pki/specs/pki-core.spec
@@ -17,6 +17,7 @@
 BuildRequires:    cmake
 BuildRequires:    java-devel >= 1:1.6.0
 BuildRequires:    ldapjdk
+BuildRequires:    apache-commons-codec
 BuildRequires:    nspr-devel
 BuildRequires:    nss-devel
 BuildRequires:    openldap-devel
@@ -185,6 +186,7 @@
 
 Requires:         java >= 1:1.6.0
 Requires:         ldapjdk
+Requires:         apache-commons-codec
 %if 0%{?fedora} >= 16
 Requires:         jpackage-utils >= 0:1.7.5-10
 Requires:         jss >= 4.2.6-19.1
--- a/pki/tools/jar/README.jar-tools
+++ b/pki/tools/jar/README.jar-tools
@@ -225,9 +225,10 @@
 The summary is listed below:
 
 Summary:
-21 Unique Jar's
+22 Unique Jar's
      /usr/lib/jss/jss4-4.2.6.jar
      /usr/lib/osutil/osutil-2.0.0.jar
+     /usr/share/java/commons-codec.jar
      /usr/lib/symkey/symkey-9.0.0.jar
      /usr/share/java/jakarta-taglibs-core-1.1.1.jar
      /usr/share/java/ldapbeans-4.18.jar
@@ -247,11 +248,12 @@
      /usr/share/java/xerces-j2-2.9.0.jar
      /usr/share/java/xml-commons-apis-1.4.01.jar
      /usr/share/java/xml-commons-apis-ext-1.4.01.jar
-15 Unique RPM's
+16 Unique RPM's
      jakarta-taglibs-standard
      jss
      ldapjdk
      osutil
+     apache-commons-codec
      pki-common
      pki-console
      pki-symkey
@@ -264,13 +266,13 @@
      xerces-j2
      xml-commons-apis
 
-What this is telling us is that there are 21 jars which provide all
+What this is telling us is that there are 22 jars which provide all
 the classes needed to satisfy the import statements. However there may
 be some duplicates in the list. Also because of wildcard import
 statements some classes and hence jars may have been included which
 are not actually utilized in the code.
 
-Those 21 jars are provided by the 15 RPM's listed. Once again there
+Those 22 jars are provided by the 16 RPM's listed. Once again there
 may be some class duplication and/or unnecessary RPM's due to
 wildcarding. But this gives a very small manageable list to manually
 pick through and make our choices. For example, one thing we can
